### PR TITLE
501: Manual Imputation Trimming

### PR DESCRIPTION
**Pull request by jwestw**
## Pull Request submission

*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `environment.yml`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `environment.yaml`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/imputation/imputation_main.py [hunk 1]
```diff
@@ -53,13 +58,28 @@ def run_imputation(
     for col in orig_cols:
         df[f"{col}_imputed"] = df[col]
 
+    # Create imp_path variable for QA output and manual imputation file
+    NETWORK_OR_HDFS = config["global"]["network_or_hdfs"]
+    imp_path = config[f"{NETWORK_OR_HDFS}_paths"]["imputation_path"]
+
+    # Load manual imputation file
+    df, manually_trimmed_df = mimp.merge_manual_imputation(
+        df, config, is_file, read_csv, imp_path
```
Comment (AnneONS) [inline src/imputation/imputation_main.py]: I can see you're in the middle of updating this: you don't need to read in the csv any more as already done

File: src/imputation/MoR.py [hunk 2]
```diff
@@ -50,6 +58,16 @@ def mor_preprocessing(df, backdata):
...
     """
+
+    # Catch dataframes which are missing cols 200, 201, formtype, status or manual_trim
+    needed_cols = {"200", "201", "formtype", "status", "manual_trim"}
+    if not needed_cols.issubset(df.columns):
+        missing = needed_cols - set(df.columns)
+        raise ValueError(
+            "The DataFrame must contain columns 200, 201, formtype, status and manual_trim."  # noqa
+            "Missing: {}".format("\n - ".join(missing))
+        )
```
Comment (AnneONS) [inline src/imputation/MoR.py]: Do we still need this in here? I don't think we use manual_trim in MoR anymore?

File: src/imputation/tmi_imputation.py [hunk 3]
```diff
@@ -560,10 +560,17 @@ def run_tmi(
...
 
     # create logic to select rows for longform and shortform TMI
-    long_tmi_mask = (full_df["formtype"] == formtype_long) & ~mor_mask
-    short_tmi_mask = (full_df["formtype"] == formtype_short) & ~mor_mask
+    long_tmi_mask = ((full_df["formtype"] == formtype_long) 
+                     & ~mor_mask 
+                     & ~manual_trim_mask)
+    short_tmi_mask = ((full_df["formtype"] == formtype_short) 
+                      & ~mor_mask 
+                      & ~manual_trim_mask)
```
Comment (AnneONS) [inline src/imputation/tmi_imputation.py]: do we need this in here? Haven't we already excluded the rows with manual_trim = True?

**Review by AnneONS — COMMENTED** [review]
I can see you're still updating this: I've put in a few comments in case useful.

I don't think you'll need to do anything with manual trimming in TMI or MoR as you're excluding the relevant rows in imputation_main.py

File: src/imputation/MoR.py (hunk 2 above)
Comment (jwestw) [reply src/imputation/MoR.py]: I think this doesn't hurt? Maybe we don't need to check for manual_trim

Can be removed if we are guaranteed that the df will always have these columns, hence making this redundant code.

File: src/staging/staging_main.py [hunk 4]
```diff
@@ -397,6 +399,28 @@ def run_staging(
         manual_outliers = None
         StagingMainLogger.info("Loading of Manual Outlier File skipped")
 
+    # Get the latest manual trim file
+    manual_trim_path = paths["manual_imp_trim_path"]
+    latest_man_file = list_files(manual_trim_path, ext="csv", order="newest")[0]
```
Comment (dpurches) [inline src/staging/staging_main.py]: I think this is nice, but would probs be safer (and easier) just directly specify which trimming file to use.

File: src/imputation/tmi_imputation.py (hunk 3 above)
Comment (AnneONS) [reply src/imputation/tmi_imputation.py]: @dpurches we don't need to do anything with the manual_trim in TMI (or MoR) as we've excluded those rows before imputation.

File: src/imputation/tmi_imputation.py [hunk 5]
```diff
@@ -560,10 +560,17 @@ def run_tmi(
     # logic to identify rows that have had MoR or CF applied,
     # these should be excluded from TMI
     mor_mask = full_df["imp_marker"].isin(["CF", "MoR"])
+    
+    # create mask to exlude records that have been manually trimmed
+    manual_trim_mask = full_df["manual_trim"] # True if record has been manually trimmed
```
Comment (AnneONS) [inline src/imputation/tmi_imputation.py]: not needed

**Review by AnneONS — COMMENTED** [review]
Just to note that we don't need to worry about excluding the manual_trim records in TMI as we already exclude them before imputation

File: src/developer_config.yaml [hunk 6]
```diff
@@ -81,6 +83,7 @@ network_paths:
   construction_amend_path:  "R:/BERD Results System Development 2023/DAP_emulation/construction/manual_construction/construction_amendments_2023-10-31_v2.csv" # TODO Need to test
   estimation_path:  "R:/BERD Results System Development 2023/DAP_emulation/estimation"
   imputation_path:  "R:/BERD Results System Development 2023/DAP_emulation/imputation"
+  manual_imp_trim_path: "R:/BERD Results System Development 2023/DAP_emulation/imputation/manual_trimming/trimming_qa_2023-11-27_v359.csv"
```
Comment (jwestw) [inline src/developer_config.yaml:86]: I notice here there is a difference between the hdfs path and the network one as the specific csv file is specified in the network one, suggesting that they work differently.

File: src/utils/hdfs_mods.py [hunk 7]
```diff
@@ -320,28 +323,28 @@ def hdfs_move_file(src_path: str, dst_path: str):
...
-def hdfs_list_files(path: str, ext: str = None):
+def hdfs_list_files(path: str, ext: str = None, order=None):
     """
     List files in a directory. Uses 'hadoop fs -ls'.
     """
     # Forming the command line command and executing
     command = ["hadoop", "fs", "-ls", path]
+    if order:
+        ord_dict = {"newest": "-t", "oldest": "-t -r"}
+        command = ["hadoop", "fs", "-ls", ord_dict[order], path]
```
Comment (jwestw) [inline src/utils/hdfs_mods.py:334]: I think ditch this code if we are not using it.

**Review by jwestw — COMMENTED** [review]
So far so good.

**Comment by dpurches** [comment]
Merged as part of #147

**Pull request closed without merging**